Implement windows agent, and some refactoring

Yukinobu-Mine opened the pull request:
*Issue #, if available:*

*Description of changes:*
- Rename `AgentLinux` to `AgentEC2Fleet`
- Implement Windows Server 2019 agent.
- Add environment variables to Jenkins server, and refer to them from `initialConfig` and `jenkins.yaml`.
  - SSH_CREDENTIALS_ID_UNIX
  - SSH_CREDENTIALS_ID_WINDOWS
- Add SSH connection parameters to `AgentEC2Fleet`.
- Add 'prefixStartSlaveCmd' and 'suffixStartSlaveCmd' to EC2 fleet agent parameters.
- Add 'Name' tag to data volumes.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Yukinobu-Mine committed:
Fix: 'LinuxAgentProps.maxSize' of small Linux agent

Yukinobu-Mine committed:
Add Windows Server 2019 agent as EC2 fleet agnet
- Refactor 'AgentLinux' as the common implementation 'AgentEC2Fleet'

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet-linux.ts:
```diff
@@ -0,0 +1,51 @@
+    : ec2.MachineImage.latestAmazonLinux2023();
+  }
+
+  protected getRootVolumeDeviceName() {
```
getXX系メソッドの代わりにaccessorを使えますかね？
https://www.typescriptlang.org/docs/handbook/classes.html#accessors

Yukinobu-Mine replied:
引数がなければ置き換え可能ですが、実装変更によって引数がつくかも知れません。検討してみます。

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet-linux.ts:
```diff
@@ -0,0 +1,51 @@
+  constructor(scope: Construct, id: string, props: AgentEC2FleetLinuxProps) {
+    super(scope, id, props);
+
+    this.amiId = props.amiId;
```
これはsuperを呼んだ後の代入で大丈夫ですか？

Yukinobu-Mine replied:
確かにそうですね 🤔 修正します。

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet-linux.ts:
```diff
@@ -0,0 +1,51 @@
+  /**
+   * @default the latest Amazon Linux 2 image.
+   */
+  readonly amiId?: string;
```
これはAgentEC2FleetPropsに入れても良い？

Yukinobu-Mine replied:
その他のプロパティも含めて AgentEC2FleetProps にまとめる選択肢もありですね。

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet-linux.ts:
```diff
@@ -0,0 +1,51 @@
+  public readonly amiId?: string;
+  public readonly credentialsId: string;
+  public readonly fsRoot: string;
+  public readonly javaPath?: string;
```
このあたりのプロパティはAgentEC2Fleetから引き継げないでしょうか

Yukinobu-Mine replied:
AgentEC2Fleet 内の実装で直接使っていないので派生クラスに置きましたが、プロパティを AgentEC2FleetProps にまとめる場合は、こちらも AgentEC2Fleet に集約する事になりますね。

tmokmss replied:
AgentEC2Fleetが共通にもっているべきプロパティだと思うので、AgentEC2Fleetに移すのが正しそうですね

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet-windows.ts:
```diff
@@ -0,0 +1,49 @@
+
+  protected getMachineImage() {
+    return ec2.MachineImage.genericWindows({
+      [cdk.Stack.of(this).region]: this.amiId ?? 'ami-0a249b7e15c3c080e', // EC2LaunchV2-Windows_Server-2019-English-Full-ContainersLatest-2023.04.12
```
これを使えますか？
https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.WindowsImage.html

Yukinobu-Mine replied:
その一覧にはない Microsoft 製の Community AMI なので ID 直書きになってますが、変わる可能性もあります。

tmokmss replied:
なるほど！Community AMIを使う理由はなんですか？

Yukinobu-Mine replied:
以下の条件を満たす AMI が Community にしかなかったです。
- Windows Server 2019
- EC2LaunchV2 対応
- docker 環境構築済み

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet-windows.ts:
```diff
@@ -0,0 +1,49 @@
+    this.amiId = props.amiId;
+    this.credentialsId = 'instance-ssh-key-administrator';
+    this.fsRoot = 'C:\\Jenkins';
+    this.javaPath = 'C:\\Java\\jdk17.0.7_7\\bin\\java.exe';
```
getXx系メソッドとプロパティはどう使い分けていますか？

Yukinobu-Mine replied:
AgentEC2Fleet クラスの内部実装で必要とするが、派生クラスで定義したいデータを abstract の getXx にしています。

tmokmss commented on lib/construct/jenkins/resources/config/initialConfig.groovy:
```diff
@@ -14,13 +14,22 @@ import jenkins.model.JenkinsLocationConfiguration
+BasicSSHUserPrivateKey ec2UserCredentials = new BasicSSHUserPrivateKey(
   CredentialsScope.GLOBAL,
-  'instance-ssh-key',
+  'instance-ssh-key-ec2-user',
```
シークレット名は環境変数で外から渡して、二重管理をやめた方が良いかも

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet-windows.ts:
```diff
@@ -0,0 +1,49 @@
+    this.amiId = props.amiId;
+    this.credentialsId = 'instance-ssh-key-administrator';
+    this.fsRoot = 'C:\\Jenkins';
+    this.javaPath = 'C:\\Java\\jdk17.0.7_7\\bin\\java.exe';
```
このJavaのパスはjavaのバージョン固定していれば不変ですかね？ `_7` というのが何なのか気になる 👀

Yukinobu-Mine replied:
`_7` はパッチ番号か何かだと思いますが、Windows 向け User Data では 17.0.7 の部分までしか指定できないので、もし `17.0.7_8` とかが出たら変更が必要になると思います...

tmokmss replied:
謎番号…
ちなみにSSH ConnectorはJavaパスの発見機能があるようですが、これはWindowsだと動かないやつですか？

Yukinobu-Mine replied:
はい、動かないです 😞 
あと、OpenSSH でログインすると、パスとかも通ってない状態になります。

tmokmss replied:
メモ:  javaPathはなくても動くことが分かったので、削除した

tmokmss reviewed (commented):
素晴らしいPRをありがとうございます！細かい点をいくつかコメントしましたー

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet-windows.ts:
```diff
@@ -0,0 +1,49 @@
+
+  protected getMachineImage() {
+    return ec2.MachineImage.genericWindows({
+      [cdk.Stack.of(this).region]: this.amiId ?? 'ami-0a249b7e15c3c080e', // EC2LaunchV2-Windows_Server-2019-English-Full-ContainersLatest-2023.04.12
```
`ami-0a249b7e15c3c080e` は us-east-2 限定だと思うので、どうせならこんな感じにしたほうが親切？
今だとus-east-2以外のリージョンでも `ami-0a249b7e15c3c080e` が使われてしまい、デプロイ時にエラーになります
↓のようにすると、AMIを書かれていない他のリージョンではsynth時にエラーになります

```ts
  protected getMachineImage() {
    return ec2.MachineImage.genericWindows({
      // EC2LaunchV2-Windows_Server-2019-English-Full-ContainersLatest-2023.04.12
      'us-east-2': 'ami-0a249b7e15c3c080e',
      'us-east-1': 'ami-080549836b0b86c26',
      ...(this.amiId
        ? {
            [cdk.Stack.of(this).region]: this.amiId,
          }
        : {}),
    });
  }
```

Yukinobu-Mine replied:
さらに調べたところ、`EC2LaunchV2-Windows_Server-2019-English-Full-ContainersLatest` は `WindowsVersion` には含まれていませんが、AWS が管理する AMI であり、以下の形で AMI ID を参照できるようです。
```
ec2.MachineImage.fromSsmParameter('/aws/service/ami-windows-latest/EC2LaunchV2-Windows_Server-2019-English-Full-ContainersLatest', {
  os: ec2.OperatingSystemType.WINDOWS,
});
```

Yukinobu-Mine committed:
Reflect review comments in pull request #3
- Reimplement 'AgentEC2Fleet'.
- Add environment variable to Jenkins server, and reference it from 'initialConfig' and 'jenkins.yaml'.
  - CREDENTIALS_ID_EC2_USER
  - CREDENTIALS_ID_ADMINISTRATOR
- Add SSH connection parameters to 'AgentEC2Fleet' and 'AgentMac'
- Remove 'javaPath' from EC2 fleet parameters.

Yukinobu-Mine committed:
Aggregate EC2 fleet agents into an array 'ec2FleetAgents'

Yukinobu-Mine committed:
Implement feature of data volume attachment to Windows agent.
- Change user data of Windows agent.
- Add 'prefixStartSlaveCmd' and 'suffixStartSlaveCmd' to agent parameters.
- Add default SSH connection parameters to 'AgentEC2FleetFactory'.
- Allow not to specify SSH connection parameters.

Yukinobu-Mine committed:
Add 'Name' tag to data volumes.

Yukinobu-Mine committed:
Merge branch 'main' into implement-windows-agent

tmokmss commented on lib/construct/jenkins/resources/config/initialConfig.groovy:
```diff
@@ -12,15 +12,27 @@ import jenkins.model.JenkinsLocationConfiguration
+BasicSSHUserPrivateKey ec2UserCredentials = new BasicSSHUserPrivateKey(
   CredentialsScope.GLOBAL,
-  'instance-ssh-key',
+  ec2UserCredentialsId,
```
`unixSshCredentialId`, `windowsSshCredentialId` はどうでしょう　(他の環境変数などもすべて

mac/linuxも分離して良いかもしれないですが…

Yukinobu-Mine replied:
とりあえずご提案の名前に変更しました。

tmokmss commented on lib/construct/jenkins/resources/agent-userdata-windows.yml:
```diff
@@ -0,0 +1,58 @@
+      $INSTANCE_ID = $instanceMetadata[1]
+      $REGION = $instanceMetadata[2].SystemName
+
+      $VOLUME_ID = Get-EC2Volume -Filter @(@{Name="tag:Kind"; Values="<KIND_TAG>"}, @{Name="availability-zone"; Values=$AZ}, @{Name="status"; Values="available"}) -Region $REGION -Select "Volumes.VolumeId" | Select-Object -First 1
```
Windowsだとこんな書き方なんですね、独特だ… https://aws.amazon.com/powershell/

tmokmss commented on lib/construct/jenkins/resources/agent-userdata-windows.yml:
```diff
@@ -0,0 +1,58 @@
+            Write-Output "drive $JENKINS_DRIVE already mounted"
+          } else {
+            Write-Output "change drive letter of first partition to $JENKINS_DRIVE"
+            $partitions | Select-Object -First 1 | Set-Partition -NewDriveLetter $JENKINS_DRIVE
```
このボリュームはインスタンス再起動時も保持されますか？(念の為確認です)

Yukinobu-Mine replied:
保持されているようです

tmokmss commented on lib/construct/jenkins/resources/config/jenkins.yaml.ejs:
```diff
@@ -52,22 +52,34 @@ jenkins:
+] : []),
+...(agent.sshConnectRetryWaitTime != null ? [
+`            retryWaitTime: ${agent.sshConnectRetryWaitTime}`,
+] : []),
```
nullableにする必要ありますかね？CDK側でデフォルト値を入れるのはどうでしょう

Yukinobu-Mine replied:
ssh プラグイン自体が持つデフォルト値は使わない方向ですか？

tmokmss replied:
はい、CDK側で適当なデフォルト値を入れていればよいかと

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet.ts:
```diff
@@ -0,0 +1,212 @@
+  readonly policyStatements?: iam.PolicyStatement[];
+}
+
+export interface AgentEC2FleetFactory {
```
あえてこのインターフェースを用意する意味は何でしょうか？単にpropsで渡すのではだめなんでしょうか

Yukinobu-Mine replied:
`AgentEC2FleetLinuxProps` が `AgentEC2FleetProps` を継承しているので、継承元に必須プロパティを追加すると、呼び出し側で指定する必要が生じてしまうためです。
`AgentEC2Fleet.linuxFleet()` 方式なら事情が少し変わるので、もう少し検討してみます。

Yukinobu-Mine replied:
やりたい事としては
- `AgentEC2Fleet` のコンストラクタでは、machineImage や userData の指定が必須
- Windows や Linux 向けでは、それらの指定が不要
- その他の props は共通 interface で定義

という動作です。`AgentEC2FleetPropsBase` みたいなのを作って、そこから `AgentEC2FleetProps` `AgentEC2FleetWindowsProps` `AgentEC2FleetLinuxProps` を派生させる方法を思いついたので、`AgentEC2Fleet.linuxFleet(scope, id, props)` パターンへの変更と併せて対応します。

tmokmss replied:
Windows や Linux 向けではmachineImage や userData の指定が不要というのは、直感に反していますが、正しいですか？

Yukinobu-Mine replied:
`AgentEC2Fleet.linuxFleet()` 方式に変更したので、詳細はそちらをご確認下さい。

tmokmss commented on lib/construct/jenkins/resources/agent-userdata-windows.yml:
```diff
@@ -0,0 +1,58 @@
+    type: powershell
+    runAs: admin
+    content: |-
+      $JENKINS_DRIVE = "D"
```
Volumeプールを利用しないときも、Dドライブはマウントされますか？

tmokmss commented on lib/construct/jenkins/resources/agent-userdata-windows.yml:
```diff
@@ -0,0 +1,58 @@
+
+      Set-ExecutionPolicy Bypass -Scope Process -Force
+      [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072
+      iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))
```
このコマンドがなぜ必要かのコメントがほしいです　(ぱっと見でわからないため)

tmokmss replied:
こちら未対応のようです　https://github.com/aws-samples/jenkins-unity-build-on-aws/pull/3#discussion_r1217337540

Yukinobu-Mine replied:
上にあるリンク先に書かれた、chocolatey のインストール用のコマンドそのままです。
リンク先では1行にまとめて書かれているのを分割していますが、元のままコピーした方が意図が伝わりやすいでしょうか？

tmokmss replied:
なるほど！であればこのままでOKです 🙆‍♂️

tmokmss commented on lib/construct/jenkins/resources/config/jenkins.yaml.ejs:
```diff
@@ -27,23 +27,23 @@ jenkins:
 [
 `    - permanent:`,
-`        labelString: mac`,
+`        labelString: '${agent.label}'`,
```
ここは一旦固定値で良いと思います　今のところmacインスタンスにバリエーションをもたせるユースケースはないので

tmokmss commented on lib/construct/jenkins/resources/config/jenkins.yaml.ejs:
```diff
@@ -27,23 +27,23 @@ jenkins:
+`            credentialsId: \${${agent.credentialsIdEnv}}`,
+`            launchTimeoutSeconds: ${agent.sshConnectTimeoutSeconds}`,
+`            maxNumRetries: ${agent.sshConnectMaxNumRetries}`,
+`            retryWaitTime: ${agent.sshConnectRetryWaitTime}`,
```
この3つも固定値で良いと思います

いまのところmacインスタンスごとにこれらの値を変えるべき状況は考えられないので

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet.ts:
```diff
@@ -0,0 +1,212 @@
+
+    this.sshConnectTimeoutSeconds = props.sshConnectTimeoutSeconds ?? factory.defaultSshConnectTimeoutSeconds;
+    this.sshConnectMaxNumRetries = props.sshConnectMaxNumRetries ?? factory.defaultSshConnectMaxNumRetries;
+    this.sshConnectRetryWaitTime = props.sshConnectRetryWaitTime ?? factory.defaultSshConnectRetryWaitTime;
```
ここを二段階で設定できる必要性は何でしょう？

Yukinobu-Mine replied:
以下のような動作を実現するためですが、呼び出し元による明示的指定だけで十分ですか?
1. 呼び出し元が props で値を指定していれば、それを適用
2. そうでなければ、Windows や Linux 向けのデフォルト値を適用
3. それも指定されていなければ、ssh plugin 自体のデフォルト値を使用

tmokmss replied:
1と2で十分と思います　かつ2の値はハードコードして良いと思います

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet.ts:
```diff
@@ -0,0 +1,212 @@
+ * Fleet of EC2 instances for Jenkins agents.
+ * The number of instances is supposed to be controlled by Jenkins EC2 Fleet plugin.
+ */
+export class AgentEC2Fleet extends Construct {
```
CDKでよく見る実装パターンとしては、いわゆるFactoryパターンが多いですよね　今回も継承よりはそちらの方がわかりやすくなるのかなと思い始めました

```ts
AgentEC2Fleet.linuxFleet(scope, id, props)
AgentEC2Fleet.windowsFleet(scope, id, props) 
```
のような

どう思われますか？

tmokmss commented on lib/jenkins-unity-build-stack.ts:
```diff
@@ -160,6 +172,45 @@ export class JenkinsUnityBuildStack extends cdk.Stack {
+      policyStatements: [
+        // policy required to run detachFromAsg job.
+        new PolicyStatement({
+          actions: ['ec2:DescribeImages', 'autoscaling:DetachInstances'],
```
これ不要なはず？AMIを作るジョブに対応しているんでしたっけ

Yukinobu-Mine replied:
Linux fleet の設定をそのまま持ってきただけでした。`detachFromAsg` に対応しなければ不要そうなので削除します。

tmokmss reviewed (commented):
取り急ぎ気づいた点をコメントしました、ありがとうございます！

Yukinobu-Mine commented:
@tmokmss 一通り返信しました。以下の修正を行います。
- `AgentMac` の credentialsId 以外の設定を元に戻す
- `AgentEC2FleetWindows` と `AgentEC2FleetLinux` を廃止し、factory method 化
- `AgentEC2FleetWindows` への policyStatements 指定を削除
- jenkins.yaml.ejs の sshConnectTimeoutSeconds 等を non nullable に戻す
- credentialsId 向け環境変数の名前を変更

Yukinobu-Mine committed:
Reflect review comments in pull request #3
- `AgentMac`
  - Remove parameters
    - label
    - sshConnectTimeoutSeconds
    - sshConnectMaxNumRetries
    - sshConnectRetryWaitTime
  - Change parameter name
    - credentialsIdEnv -> sshCredentialsIdEnv
- `AgentEC2Fleet`
  - Make parameters not optional
    - fsRoot
    - sshConnectTimeoutSeconds
    - sshConnectMaxNumRetries
    - sshConnectRetryWaitTime
  - Change parameter name
    - credentialsIdEnv -> sshCredentialsIdEnv
- `AgentEC2FleetLinux` and `AgentEC2FleetWindows`
  - Remove and implement as factory method of `AgentEC2Fleet`
- Remove policyStatements of `AgentEC2FleetWindows`
  - 'ec2:DescribeImages'
  - 'autoscaling:DetachInstances'
- Change name of Jenkins environment variables
  - CREDENTIALS_ID_EC2_USER -> SSH_CREDENTIALS_ID_UNIX
  - CREDENTIALS_ID_ADMINISTRATOR -> SSH_CREDENTIALS_ID_WINDOWS

Yukinobu-Mine commented:
@tmokmss 以下の修正を行いました
- `AgentMac` の credentialsId 以外の設定を元に戻す
- `AgentEC2FleetWindows` と `AgentEC2FleetLinux` を廃止し、factory method 化
- `AgentEC2FleetWindows` への policyStatements 指定を削除
- jenkins.yaml.ejs の sshConnectTimeoutSeconds 等を non nullable に戻す
- credentialsId 向け環境変数の名前を変更

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet.ts:
```diff
@@ -0,0 +1,259 @@
+      }),
+      userData: userData,
+      rootVolumeDeviceName: '/dev/sda1',
+      fsRoot: 'D:\\Jenkins',
```
Dドライブはボリュームプール使わない場合も使えますか？

Yukinobu-Mine replied:
使えないので、windows版のsmallを作る場合は fsRoot を変える必要があります。

tmokmss replied:
お願いします！dataVolumeSize == nullで分岐すればOKです

tmokmss commented on lib/construct/jenkins/agent-ec2-fleet.ts:
```diff
@@ -0,0 +1,259 @@
+    return new AgentEC2Fleet(scope, id, {
+      machineImage: props.amiId
+      ? ec2.MachineImage.genericWindows({ [cdk.Stack.of(scope).region]: props.amiId })
+      : ec2.MachineImage.fromSsmParameter('/aws/service/ami-windows-latest/EC2LaunchV2-Windows_Server-2019-English-Full-ContainersLatest', {
```
このSSM Parameter名が明記されたドキュメントへのリンクを書いておいてくださいmm

tmokmss replied:
ちなみにWin server 2022もあるようですが、2019にしたのは意図がありますか？

Yukinobu-Mine replied:
gameci にある Unity の Windows 向け Docker イメージが 2019 向けだからです。
EC2 では、異なるバージョンのWindowsカーネルを使う仮想化レベルはベアメタルでしか使えないようです。
https://kkamegawa.hatenablog.jp/entry/2021/03/16/205421

tmokmss replied:
なるほどです、それもコメントに書いておきましょうー

tmokmss commented on lib/construct/jenkins/controller.ts:
```diff
@@ -77,18 +93,23 @@ export class Controller extends Construct {
+
+    const fleetLaunchTemplateIdEnv = (agent: { name: string }) =>
       `FLEET_LAUNCH_TEMPLATE_ID_${agent.name.toUpperCase().replace(/-/g, '_')}`;
-    const macEnv = (agent: { name: string }) => `MAC_HOST_${agent.name.toUpperCase().replace(/-/g, '_')}`;
```
幅120文字まで許容しているはずなのですが、prettierrcは読み込めていますかね？
https://github.com/aws-samples/jenkins-unity-build-on-aws/blob/main/.prettierrc

tmokmss reviewed (commented):
遅くなってすみません、ありがとうございます　ほぼほぼOKです！
Windowsはデータボリュームの分岐が必要になるかもですね

Yukinobu-Mine committed:
Reflect review comments in pull request #3
- Describe about 'EC2LaunchV2-Windows_Server-2019-English-Full-ContainersLatest'.
- Switch 'fsRoot' and 'prefixStartSlaveCmd' based on 'dataVolumeSize'.
- Reformat code by prettier.

tmokmss commented:
LGTMです！マージしておきますね 👍
